File: Cyaim.WebSocketServer/Clients/Cyaim.WebSocketServer.Client/WebSocketClientFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Reflection;
using System.Text.Json;
using System.Threading.Tasks;

namespace Cyaim.WebSocketServer.Client
{
    /// <summary>
    /// Factory for creating WebSocket client proxies based on server endpoints
    /// 基于服务器端点创建 WebSocket 客户端代理的工厂
    /// </summary>
    public class WebSocketClientFactory
    {
        private readonly string _serverBaseUrl;
        private readonly string _channel;
        private readonly HttpClient _httpClient;
        private List<WebSocketEndpointInfo>? _cachedEndpoints;
        private readonly WebSocketClientOptions _options;

        /// <summary>
        /// Constructor / 构造函数
        /// </summary>
        /// <param name="serverBaseUrl">Server base URL (e.g., "http://localhost:5000") / 服务器基础 URL（例如："http://localhost:5000"）</param>
        /// <param name="channel">WebSocket channel (default: "/ws") / WebSocket 通道（默认："/ws"）</param>
        /// <param name="options">Client creation options / 客户端创建选项</param>
        public WebSocketClientFactory(string serverBaseUrl, string channel = "/ws", WebSocketClientOptions? options = null)
        {
            _serverBaseUrl = serverBaseUrl?.TrimEnd('/') ?? throw new ArgumentNullException(nameof(serverBaseUrl));
            _channel = channel ?? throw new ArgumentNullException(nameof(channel));
            _options = options ?? new WebSocketClientOptions();
            _httpClient = new HttpClient();
        }

        /// <summary>
        /// Get endpoints from server / 从服务器获取端点
        /// </summary>
        public async Task<List<WebSocketEndpointInfo>> GetEndpointsAsync()
        {
            if (_cachedEndpoints != null)
            {
                return _cachedEndpoints;
            }

            try
            {
                var url = $"{_serverBaseUrl}/ws_server/api/endpoints";
                var response = await _httpClient.GetAsync(url);
                response.EnsureSuccessStatusCode();

                var json = await response.Content.ReadAsStringAsync();
                var apiResponse = JsonSerializer.Deserialize<ApiResponse<List<WebSocketEndpointInfo>>>(
                    json,
                    new JsonSerializerOptions { PropertyNameCaseInsensitive = true }
                );

                if (apiResponse?.Success == true && apiResponse.Data != null)
                {
                    _cachedEndpoints = apiResponse.Data;
                    return _cachedEndpoints;
                }

                throw new Exception(apiResponse?.Error ?? "Failed to fetch endpoints");
            }
            catch (Exception ex)
            {
                throw new Exception($"Error fetching WebSocket endpoints from server: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Create a client proxy for a specific interface / 为特定接口创建客户端代理
        /// </summary>
        /// <typeparam name="T">Interface type / 接口类型</typeparam>
        /// <returns>Client proxy instance / 客户端代理实例</returns>
        public async Task<T> CreateClientAsync<T>() where T : class
        {
            var interfaceType = typeof(T);

            if (!interfaceType.IsInterface)
            {
                throw new ArgumentException($"Type {interfaceType.Name} must be an interface");
            }

            // Create WebSocket client / 创建 WebSocket 客户端
            var wsUri = _serverBaseUrl.Replace("http://", "ws://").Replace("https://", "wss://");
            var client = new WebSocketClient(wsUri, _channel, _options);

            // Get endpoints if not lazy loading / 如果不是延迟加载，则获取端点
            List<WebSocketEndpointInfo>? endpoints = null;
            if (!_options.LazyLoadEndpoints)
            {
                endpoints = await GetEndpointsAsync();
            }

            // Create proxy using reflection / 使用反射创建代理
            return CreateProxy<T>(client, endpoints);
        }

        /// <summary>
        /// Create proxy instance / 创建代理实例
        /// </summary>
        private T CreateProxy<T>(WebSocketClient client, List<WebSocketEndpointInfo>? endpoints) where T : class
        {
            var interfaceType = typeof(T);
            var methods = interfaceType.GetMethods();

            // Create a dictionary to map method names to endpoints / 创建方法名到端点的映射
            var methodEndpointMap = new Dictionary<MethodInfo, WebSocketEndpointInfo?>();
            var missingMethods = new List<string>();

            foreach (var method in methods)
            {
                WebSocketEndpointInfo? endpoint = null;

                // First, check for WebSocketEndpointAttribute / 首先检查 WebSocketEndpointAttribute
                var endpointAttr = method.GetCustomAttribute<WebSocketEndpointAttribute>();
                if (endpointAttr != null)
                {
                    // If endpoints are available, validate the target / 如果端点可用，验证目标
                    if (endpoints != null)
                    {
                        endpoint = endpoints.FirstOrDefault(ep =>
                            ep.Target.Equals(endpointAttr.Target, StringComparison.OrdinalIgnoreCase) ||
                            ep.MethodPath.Equals(endpointAttr.Target, StringComparison.OrdinalIgnoreCase) ||
                            ep.FullName.Equals(endpointAttr.Target, StringComparison.OrdinalIgnoreCase));
                    }
                    else
                    {
                        // If lazy loading, store the target from attribute / 如果延迟加载，存储特性中的目标
                        endpoint = new WebSocketEndpointInfo { Target = endpointAttr.Target };
                    }
                }
                else if (endpoints != null)
                {
                    // Try to find matching endpoint by method name / 尝试通过方法名查找匹配的端点
                    endpoint = endpoints.FirstOrDefault(ep =>
                        ep.Action.Equals(method.Name, StringComparison.OrdinalIgnoreCase) ||
                        ep.MethodPath.Equals(method.Name, StringComparison.OrdinalIgnoreCase));
                }

                if (endpoint != null)
                {
                    methodEndpointMap[method] = endpoint;
                }
                else
                {
                    methodEndpointMap[method] = null;
                    if (_options.ValidateAllMethods)
                    {
                        missingMethods.Add(method.Name);
                    }
                }
            }

            // Validate all methods if required / 如果需要，验证所有方法
            if (_options.ValidateAllMethods && missingMethods.Count > 0)
            {
                throw new InvalidOperationException(
                    $"The following methods do not have corresponding endpoints: {string.Join(", ", missingMethods)}. " +
                    $"You can use [WebSocketEndpoint(\"target\")] attribute to specify the endpoint, or set ValidateAllMethods to false.");
            }

            // Create proxy using DispatchProxy / 使用 DispatchProxy 创建代理
            var proxy = DispatchProxy.Create<T, WebSocketClientProxy<T>>();
            if (proxy is WebSocketClientProxy<T> typedProxy)
            {
                typedProxy.Initialize(client, methodEndpointMap, this, _options);
            }

            return proxy;
        }

        /// <summary>
        /// Dispose resources / 释放资源
        /// </summary>
        public void Dispose()
        {
            _httpClient?.Dispose();
        }
    }

    /// <summary>
    /// WebSocket client proxy implementation / WebSocket 客户端代理实现
    /// </summary>
    internal class WebSocketClientProxy<T> : DispatchProxy where T : class
    {
        private WebSocketClient? _client;
        private Dictionary<MethodInfo, WebSocketEndpointInfo?>? _methodEndpointMap;
        private WebSocketClientFactory? _factory;
        private WebSocketClientOptions? _options;

        /// <summary>
        /// Initialize proxy / 初始化代理
        /// </summary>
        public void Initialize(
            WebSocketClient client,
            Dictionary<MethodInfo, WebSocketEndpointInfo?> methodEndpointMap,
            WebSocketClientFactory factory,
            WebSocketClientOptions options)
        {
            _client = client;
            _methodEndpointMap = methodEndpointMap;
            _factory = factory;
            _options = options;
        }

        protected override object? Invoke(MethodInfo? targetMethod, object?[]? args)
        {
            if (targetMethod == null || _client == null || _methodEndpointMap == null || _factory == null || _options == null)
            {
                throw new InvalidOperationException("Proxy not properly initialized");
            }

            // Get endpoint, lazy load if needed / 获取端点，如果需要则延迟加载
            if (!_methodEndpointMap.TryGetValue(targetMethod, out var endpoint) || endpoint == null)
            {
                // Try lazy loading / 尝试延迟加载
                if (_options.LazyLoadEndpoints)
                {
                    endpoint = LoadEndpointForMethodAsync(targetMethod).GetAwaiter().GetResult();
                    if (endpoint != null)
                    {
                        _methodEndpointMap[targetMethod] = endpoint;
                    }
                }

                if (endpoint == null)
                {
                    var errorMsg = $"Endpoint not found for method {targetMethod.Name}. " +
                        $"You can use [WebSocketEndpoint(\"target\")] attribute to specify the endpoint.";
                    
                    if (_options.ThrowOnEndpointNotFound)
                    {
                        throw new NotSupportedException(errorMsg);
                    }
                    else
                    {
                        throw new InvalidOperationException(errorMsg);
                    }
                }
            }

            // Build request body from method parameters / 从方法参数构建请求体
            object? requestBody = null;
            if (args != null && args.Length > 0)
            {
                if (args.Length == 1)
                {
                    requestBody = args[0];
                }
                else
                {
                    // Create anonymous object from parameters / 从参数创建匿名对象
                    var paramNames = targetMethod.GetParameters().Select(p => p.Name).ToArray();
                    var dict = new Dictionary<string, object?>();
                    for (int i = 0; i < args.Length && i < paramNames.Length; i++)
                    {
                        if (paramNames[i] != null)
                        {
                            dict[paramNames[i]!] = args[i];
                        }
                    }
                    requestBody = dict;
                }
            }

            // Get return type / 获取返回类型
            var returnType = targetMethod.ReturnType;
            if (returnType.IsGenericType && returnType.GetGenericTypeDefinition() == typeof(Task<>))
            {
                var resultType = returnType.GetGenericArguments()[0];
                var method = typeof(WebSocketClientProxy<T>).GetMethod(
                    nameof(InvokeAsync),
                    BindingFlags.NonPublic | BindingFlags.Instance
                )!.MakeGenericMethod(resultType);

                return method.Invoke(this, new object[] { endpoint.Target, requestBody! });
            }
            else if (returnType == typeof(Task))
            {
                return InvokeAsync<object>(endpoint.Target, requestBody!);
            }

            throw new NotSupportedException($"Unsupported return type: {returnType}");
        }

        private async Task<TResult> InvokeAsync<TResult>(string target, object requestBody)
        {
            await _client!.ConnectAsync();
            return await _client.SendRequestAsync<object, TResult>(target, requestBody);
        }

        /// <summary>
        /// Load endpoint for a method (lazy loading) / 为方法加载端点（延迟加载）
        /// </summary>
        private async Task<WebSocketEndpointInfo?> LoadEndpointForMethodAsync(MethodInfo method)
        {
            if (_factory == null)
            {
                return null;
            }

            var endpoints = await _factory.GetEndpointsAsync();

            // Check attribute first / 首先检查特性
            var endpointAttr = method.GetCustomAttribute<WebSocketEndpointAttribute>();
            if (endpointAttr != null)
            {
                return endpoints.FirstOrDefault(ep =>
                    ep.Target.Equals(endpointAttr.Target, StringComparison.OrdinalIgnoreCase) ||
                    ep.MethodPath.Equals(endpointAttr.Target, StringComparison.OrdinalIgnoreCase) ||
                    ep.FullName.Equals(endpointAttr.Target, StringComparison.OrdinalIgnoreCase));
            }

            // Try by method name / 尝试通过方法名
            return endpoints.FirstOrDefault(ep =>
                ep.Action.Equals(method.Name, StringComparison.OrdinalIgnoreCase) ||
                ep.MethodPath.Equals(method.Name, StringComparison.OrdinalIgnoreCase));
        }
    }
}

File: Cyaim.WebSocketServer/Clients/cyaim-websocket-client-js/src/WebSocketClientFactory.ts
import { WebSocketClient } from "./WebSocketClient";
import { WebSocketClientOptions } from "./WebSocketClientOptions";
import { WebSocketEndpointInfo } from "./WebSocketEndpointInfo";

interface ApiResponse<T> {
        success: boolean;
        data?: T;
        error?: string;
}

/**
 * Factory for creating WebSocket client proxies based on server endpoints
 * 基于服务器端点创建 WebSocket 客户端代理的工厂
 */
export class WebSocketClientFactory {
        private serverBaseUrl: string;
        private channel: string;
        private cachedEndpoints: WebSocketEndpointInfo[] | null = null;
        private options: WebSocketClientOptions;

        /**
         * Constructor / 构造函数
         */
        constructor(
                serverBaseUrl: string,
                channel: string = "/ws",
                options?: WebSocketClientOptions
        ) {
                this.serverBaseUrl = serverBaseUrl.replace(/\/$/, "");
                this.channel = channel;
                this.options = options || new WebSocketClientOptions();
        }

        /**
         * Get endpoints from server / 从服务器获取端点
         */
        async getEndpoints(): Promise<WebSocketEndpointInfo[]> {
                if (this.cachedEndpoints) {
                        return this.cachedEndpoints;
                }

                try {
                        const url = `${this.serverBaseUrl}/ws_server/api/endpoints`;
                        const response = await fetch(url);

                        if (!response.ok) {
                                throw new Error(
                                        `HTTP error! status: ${response.status}`
                                );
                        }

                        const apiResponse: ApiResponse<
                                WebSocketEndpointInfo[]
                        > = await response.json();

                        if (apiResponse.success && apiResponse.data) {
                                this.cachedEndpoints = apiResponse.data;
                                return this.cachedEndpoints;
                        }

                        throw new Error(
                                apiResponse.error || "Failed to fetch endpoints"
                        );
                } catch (error) {
                        throw new Error(
                                `Error fetching WebSocket endpoints from server: ${error}`
                        );
                }
        }

        /**
         * Create a client proxy for a specific interface / 为特定接口创建客户端代理
         */
        async createClient<
                T extends Record<string, (...args: any[]) => Promise<any>>
        >(interfaceDefinition: T): Promise<T> {
                const wsUri = this.serverBaseUrl.replace(/^http/, "ws");
                const client = new WebSocketClient(
                        wsUri,
                        this.channel,
                        this.options
                );

                // Get endpoints if not lazy loading / 如果不是延迟加载，则获取端点
                let endpoints: WebSocketEndpointInfo[] | null = null;
                if (!this.options.lazyLoadEndpoints) {
                        endpoints = await this.getEndpoints();
                }

                // Create proxy / 创建代理
                return this.createProxy(client, interfaceDefinition, endpoints);
        }

        /**
         * Create proxy instance / 创建代理实例
         */
        private createProxy<
                T extends Record<string, (...args: any[]) => Promise<any>>
        >(
                client: WebSocketClient,
                interfaceDefinition: T,
                endpoints: WebSocketEndpointInfo[] | null
        ): T {
                const proxy = {} as T;

                for (const [methodName, method] of Object.entries(
                        interfaceDefinition
                )) {
                        // Find endpoint for this method / 查找此方法的端点
                        let endpoint: WebSocketEndpointInfo | null = null;

                        // Check if method has endpoint metadata / 检查方法是否有端点元数据
                        const endpointTarget = (method as any).__endpointTarget;
                        if (endpointTarget) {
                                if (endpoints) {
                                        endpoint =
                                                endpoints.find(
                                                        (ep) =>
                                                                ep.target.toLowerCase() ===
                                                                        endpointTarget.toLowerCase() ||
                                                                ep.methodPath.toLowerCase() ===
                                                                        endpointTarget.toLowerCase() ||
                                                                ep.fullName.toLowerCase() ===
                                                                        endpointTarget.toLowerCase()
                                                ) || null;
                                } else {
                                        // Lazy loading: create temporary endpoint / 延迟加载：创建临时端点
                                        endpoint = {
                                                target: endpointTarget,
                                        } as WebSocketEndpointInfo;
                                }
                        } else if (endpoints) {
                                // Try to find by method name / 尝试通过方法名查找
                                endpoint =
                                        endpoints.find(
                                                (ep) =>
                                                        ep.action.toLowerCase() ===
                                                                methodName.toLowerCase() ||
                                                        ep.methodPath.toLowerCase() ===
                                                                methodName.toLowerCase()
                                        ) || null;
                        }

                        // Create proxy method / 创建代理方法
                        proxy[methodName] = async (...args: any[]) => {
                                // Lazy load endpoint if needed / 如果需要，延迟加载端点
                                if (
                                        !endpoint &&
                                        this.options.lazyLoadEndpoints
                                ) {
                                        const allEndpoints =
                                                await this.getEndpoints();
                                        const endpointTarget = (method as any)
                                                .__endpointTarget;
                                        if (endpointTarget) {
                                                endpoint =
                                                        allEndpoints.find(
                                                                (ep) =>
                                                                        ep.target.toLowerCase() ===
                                                                                endpointTarget.toLowerCase() ||
                                                                        ep.methodPath.toLowerCase() ===
                                                                                endpointTarget.toLowerCase() ||
                                                                        ep.fullName.toLowerCase() ===
                                                                                endpointTarget.toLowerCase()
                                                        ) || null;
                                        } else {
                                                endpoint =
                                                        allEndpoints.find(
                                                                (ep) =>
                                                                        ep.action.toLowerCase() ===
                                                                                methodName.toLowerCase() ||
                                                                        ep.methodPath.toLowerCase() ===
                                                                                methodName.toLowerCase()
                                                        ) || null;
                                        }
                                }

                                if (!endpoint) {
                                        const errorMsg = `Endpoint not found for method ${methodName}. Use endpoint() decorator to specify the endpoint.`;
                                        if (
                                                this.options
                                                        .throwOnEndpointNotFound
                                        ) {
                                                throw new Error(errorMsg);
                                        } else {
                                                throw new Error(errorMsg);
                                        }
                                }

                                // Build request body / 构建请求体
                                let requestBody: any = undefined;
                                if (args.length === 1) {
                                        requestBody = args[0];
                                } else if (args.length > 1) {
                                        // Create object from parameters / 从参数创建对象
                                        const paramNames =
                                                (method as any).__paramNames ||
                                                [];
                                        requestBody = {};
                                        for (
                                                let i = 0;
                                                i < args.length &&
                                                i < paramNames.length;
                                                i++
                                        ) {
                                                requestBody[paramNames[i]] =
                                                        args[i];
                                        }
                                }

                                await client.connect();
                                return await client.sendRequest<any, any>(
                                        endpoint.target,
                                        requestBody
                                );
                        };
                }

                return proxy;
        }
}

/**
 * Decorator to specify endpoint target / 用于指定端点目标的装饰器
 */
export function endpoint(target: string) {
        return function (
                targetObj: any,
                propertyKey: string,
                descriptor: PropertyDescriptor
        ) {
                descriptor.value.__endpointTarget = target;
                return descriptor;
        };
}
